src/main/scala/dpla/ingestion3/harvesters/oai/OaiRelation.scala

```diff
@@ -55,73 +56,97 @@ class OaiRelation (endpoint: String,
       }
-      case "ListRecords" => {
-        StructType(Seq(StructField("id", StringType, true),
-          StructField("document", StringType, true),
-          StructField("set_id", StringType, true),
-          StructField("set_document", StringType, true)))
+      case (None, Some(list), false) => {
+        Some(oaiResponseBuilder.getSetsExclude(list))
+      }
+      case (None, None, true) => Some(allSets)
+      case (None, None, false) => None
+      // Throw exception if more than one set handler is present.
+      case _ => {
+        val msg = "Only one of the following can be present: " +
+          "harvestAllSets, setlist, blacklist"
+        throw new IllegalArgumentException(msg)
       }
     }
   }
 
-  // Build the rows for the DataFrame.
-  override def buildScan(): RDD[Row] = {
-
+  // Get responses according to the verb.
+  def getResponses: RDD[OaiResponse] = {
     verb match {
-      case "ListSets" => {
-        allSets.map(set => Row(set.id, set.document))
-      }
-      case "ListRecords" => {
-        records.map(record => {
-          // Get set id and set document if record has associated set.
-          val (setId, setDoc) = record.set match {
-            case Some(set) => (set.id, set.document)
-            case _ => (None, None)
-          }
-          Row(record.id, record.document, setId, setDoc)
-        })
-      }
+      case "ListSets" => allSets
+      case "ListRecords" => records
       case _ => {
         val msg = s"Verb $verb not supported."
         throw new IllegalArgumentException(msg)
       }
     }
   }
+
+  /**
+    * Set the schema for the DataFrame that will be returned on load.
+    * Columns:
+    *   - set: OaiSet
+    *   - record: OaiRecord
+    *   - error: String
+    * Each row will have a non-null value for ONE of sets, records, or error.
+    */
+  override def schema: StructType = {
+    val setType = ScalaReflection.schemaFor[OaiSet].dataType
+      .asInstanceOf[StructType]
+    val recordType = ScalaReflection.schemaFor[OaiRecord].dataType
+      .asInstanceOf[StructType]
+    val errorType = ScalaReflection.schemaFor[OaiError].dataType
+      .asInstanceOf[StructType]
+
+    val setField = StructField("set", setType, true)
+    val recordField = StructField("record", recordType, true)
+    val errorField = StructField("error", errorType, true)
+    StructType(Seq(setField, recordField, errorField))
+  }
+
+  /**
+    * Build the rows for the DataFrame.
+    *
+    * Duplicate records may occur if they appeared in different sources.
+    * For example, if a record belongs to more than one set, it may appear in
+    * responses for each set.
+    */
+  override def buildScan(): RDD[Row] = {
+
+    getResponses.flatMap {
+      case x: SetsPage => x.sets.map(set => Row(set, None, None))
+      case x: RecordsPage => x.records.map(record => Row(None, record, None))
+      case x: OaiError => Seq(Row(None, None, x))
+      case x: OaiSource => {
```

Comment:
You don't need this level of curly braces just FYI. `case foo =>` statements specify their own block scopes without the braces.